### PR TITLE
[FIRRTL] Type lowering of aggregate RegResetOP

### PR DESCRIPTION
**Pull request by prithayan**
Add type lowering for RegReset op. (Fix for https://github.com/llvm/circt/issues/785)
Given  ```%name = firrtl.regreset %clockVal, %resetSignal, %resetValue : t1```
Where, ``` %name ```  and ```%resetValue``` is of aggregate type, 
create the corresponding ```firrtl.regreset ``` operation for each field type of ``` %name ```, 
and also replace ```%resetValue``` with the corresponding lowered type for corresponding field.

For example
```mlir
 %r = firrtl.regreset %clock, %reset, %init {name = "r"} : (!firrtl.clock, !firrtl.uint<1>, !firrtl.vector<uint<1>, 2>) -> !firrtl.vector<uint<1>, 2>
```
is replaced with 
```mlir
%r_0 = firrtl.regreset %clock, %reset, %init_0 {name = "r_0"} : (!firrtl.clock, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
%r_1 = firrtl.regreset %clock, %reset, %init_1 {name = "r_1"} : (!firrtl.clock, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
```

### CODE REVIEW

File: test/Dialect/FIRRTL/lower-types.mlir [hunk 1]
```diff
@@ -514,3 +514,30 @@ firrtl.circuit "ExternalModule" {
     %inst_source = firrtl.instance @ExternalModule {name = "", portNames = ["source"]} : !firrtl.bundle<valid: flip<uint<1>>, ready: uint<1>, data: flip<uint<64>>>
   }
 }
+
```
Comment (seldridge) [inline test/Dialect/FIRRTL/lower-types.mlir:517]: I think this needs the following to be picked up by `--split-input-file`?
```suggestion
// -----

```

**Review by seldridge — APPROVED** [review]
Looks great!

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp [hunk 2]
```diff
@@ -582,6 +583,38 @@ void TypeLoweringVisitor::visitDecl(RegOp op) {
...
+  // If the RegReset is not a bundle, there is nothing to do.
+  if (!resultType)
+    return;
+
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenType(resultType, "", false, fieldTypes);
+
+  // Loop over the leaf aggregates.
+  for (auto field : fieldTypes) {
+    SmallString<16> loweredName(op.nameAttr().getValue());
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:603]: I believe the name is actually an optional attribute, so you have to check if it has a value before using it.  I think the RegOp has this same issue.

Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:603]: @youngar , Thanks for catching this, confirmed it crashes when the name is not present. Will create a new issue with this

**Pull request merged**